File: android/src/main/java/com/otpcodegenerator/OtpCodeGeneratorUtils.java
package com.otpcodegenerator;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;

public class OtpCodeGeneratorUtils {
    private static final String ALGORITHM = "HmacSHA1";
    private static final String BASE32_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
    private static final int DEFAULT_PASSCODE_LENGTH = 6;
    private static final int MAX_PASSCODE_LENGTH = 9;

    public static byte[] hmacSha(byte[] keyBytes, byte[] text) throws NoSuchAlgorithmException, InvalidKeyException {
        SecretKeySpec key = new SecretKeySpec(keyBytes, ALGORITHM);
        Mac mac = Mac.getInstance(ALGORITHM);
        mac.init(key);
        return mac.doFinal(text);
    }

    public static byte[] base32Decode(String base32SecretKey) {
        base32SecretKey = base32SecretKey.toUpperCase().replaceAll(" ", ""); // Remove any spaces

        // Remove any padding characters '=' from the end of the secret key
        while (base32SecretKey.endsWith("=")) {
            base32SecretKey = base32SecretKey.substring(0, base32SecretKey.length() - 1);
        }

        int numBytes = (int) Math.ceil(base32SecretKey.length() * 5 / 8.0);
        byte[] bytes = new byte[numBytes];

        int currentByte = 0;
        int bitsRemaining = 8;
        int mask = 0;
        for (char c : base32SecretKey.toCharArray()) {
            int value = BASE32_CHARS.indexOf(c);
            if (value < 0) {
                throw new IllegalArgumentException("Invalid Base32 character: " + c);
            }
            if (bitsRemaining > 5) {
                mask = value << (bitsRemaining - 5);
                bytes[currentByte] |= mask;
                bitsRemaining -= 5;
            } else {
                mask = value >> (5 - bitsRemaining);
                bytes[currentByte] |= mask;
                currentByte++;
                bytes[currentByte] |= (value << (3 + bitsRemaining)) & 0xFF;
                bitsRemaining += 3;
            }
        }

        return bytes;
    }

    public static int truncateHash(byte[] hashValue, int passcodeLength) {
        int offset = hashValue[hashValue.length - 1] & 0xF;
        int truncatedHash = ((hashValue[offset] & 0x7F) << 24) |
                ((hashValue[offset + 1] & 0xFF) << 16) |
                ((hashValue[offset + 2] & 0xFF) << 8) |
                (hashValue[offset + 3] & 0xFF);
        return truncatedHash % (int) Math.pow(10, passcodeLength);
    }
}

File: ios/OtpCodeGeneratorUtils.swift
import Foundation

class OtpCodeGeneratorUtils {
    private static let ALGORITHM = CCHmacAlgorithm(kCCHmacAlgSHA1)
    private static let BASE32_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
    private static let DEFAULT_PASSCODE_LENGTH = 6
    private static let MAX_PASSCODE_LENGTH = 9

    static func hmacSha(key: Data, data: Data) -> Data? {
        var result = [UInt8](repeating: 0, count: Int(CC_SHA1_DIGEST_LENGTH))
        key.withUnsafeBytes { keyPtr in
            data.withUnsafeBytes { dataPtr in
                CCHmac(OtpCodeGeneratorUtils.ALGORITHM, keyPtr.baseAddress, key.count, dataPtr.baseAddress, data.count, &result)
            }
        }
        return Data(result)
    }

    static func base32Decode(_ base32SecretKey: String) -> Data? {
        let base32Chars = OtpCodeGeneratorUtils.BASE32_CHARS
        var base32SecretKey = base32SecretKey.uppercased().replacingOccurrences(of: " ", with: "") // Remove any spaces

        // Remove any padding characters '=' from the end of the secret key
        while base32SecretKey.hasSuffix("=") {
            base32SecretKey.removeLast()
        }

        var numBytes = Int(ceil(Double(base32SecretKey.count) * 5 / 8.0))
        var bytes = [UInt8](repeating: 0, count: numBytes)

        var currentByte = 0
        var bitsRemaining = 8
        var mask = 0
        for c in base32SecretKey {
            guard let value = base32Chars.firstIndex(of: c)?.encodedOffset else {
                return nil
            }

            if bitsRemaining > 5 {
                mask = value << (bitsRemaining - 5)
                bytes[currentByte] |= UInt8(mask)
                bitsRemaining -= 5
            } else {
                mask = value >> (5 - bitsRemaining)
                bytes[currentByte] |= UInt8(mask)
                currentByte += 1
                bytes[currentByte] |= UInt8((value << (3 + bitsRemaining)) & 0xFF)
                bitsRemaining += 3
            }
        }

        return Data(bytes)
    }

    static func truncateHash(hashValue: Data, passcodeLength: Int) -> Int {
        let hashArray = [UInt8](hashValue)
        let offset = Int(hashArray[hashArray.count - 1] & 0xF)

        var truncatedHash = Int(hashArray[offset] & 0x7F) << 24
        truncatedHash |= Int(hashArray[offset + 1] & 0xFF) << 16
        truncatedHash |= Int(hashArray[offset + 2] & 0xFF) << 8
        truncatedHash |= Int(hashArray[offset + 3] & 0xFF)

        return truncatedHash % Int(pow(10, Double(passcodeLength)))
    }
}
